impeller/typographer/backends/stb/text_render_context_stb.cc

```diff
@@ -0,0 +1,587 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/typographer/backends/stb/text_render_context_stb.h"
+
+#include <utility>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/trace_event.h"
+#include "impeller/base/allocation.h"
+#include "impeller/core/allocator.h"
+#include "impeller/typographer/font_glyph_pair.h"
+#include "typeface_stb.h"
+
+// These values can be customize per build.
+// Glyph atlases are always square.
+#ifndef MAX_GLYPH_ATLAS_SIZE
+#define MAX_GLYPH_ATLAS_SIZE 4096u
+#endif
+#ifndef MIN_GLYPH_ATLAS_SIZE
+#define MIN_GLYPH_ATLAS_SIZE 8u
+#endif
+
+#define DISABLE_COLOR_FONT_SUPPORT 1
+#ifdef DISABLE_COLOR_FONT_SUPPORT
+#define COLOR_FONT_BPP 1
+#else
+#define COLOR_FONT_BPP 4
+#endif
+
+namespace impeller {
+
+// An simple bitmap in lieu of a skia bitmap.
+class STBBitmap {
+ public:
+  ~STBBitmap() = default;
+
+  STBBitmap() = delete;
+
+  STBBitmap(size_t width, size_t height, size_t bytes_per_pixel)
+      : width_(width),
+        height_(height),
+        bytes_per_pixel_(bytes_per_pixel),
+        pixels_(std::make_unique<uint8_t[]>(width * height * bytes_per_pixel)) {
+  }
+
+  uint8_t* GetPixels() const { return pixels_.get(); }
+
+  uint8_t* GetPixelAddress(TPoint<size_t> coords) const {
+    FML_DCHECK(coords.x < width_);
+    FML_DCHECK(coords.x < height_);
+
+    return &pixels_.get()[(coords.x + width_ * coords.y) * bytes_per_pixel_];
+  }
+
+  size_t GetRowBytes() const { return width_ * bytes_per_pixel_; }
+
+  size_t GetWidth() const { return width_; }
+
+  size_t GetHeight() const { return height_; }
+
+  size_t GetSize() const { return width_ * height_ * bytes_per_pixel_; }
+
+ private:
+  size_t width_;
+  size_t height_;
+  size_t bytes_per_pixel_;
+  std::unique_ptr<uint8_t[]> pixels_;
+};
+
+// Analogous to the bitmaps (one for each type) stored in each Atlas context.
+static auto alpha_bitmap =
+    std::make_shared<STBBitmap>(MIN_GLYPH_ATLAS_SIZE, MIN_GLYPH_ATLAS_SIZE, 1);
+static auto color_bitmap = std::make_shared<STBBitmap>(MIN_GLYPH_ATLAS_SIZE,
```

Comment:
Nit: `gColorBitmap` per convention.

Reply:
On second thought, I don't think we have static globals elsewhere for objects as large as these. These look like leaks and on engine teardown and are hard to track down. Can we do this API right™?

Reply:
Done. Yeah, it turns out the GlyphAtlasContext still had an SkBitmap. I suspect that's the reason these were globals.

I reworked it to make backend-specific versions and got rid of the static stuff, so now both backends are storing their bitmap data in the atlas context.